pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go

```diff
@@ -0,0 +1,141 @@
+*/
+
+package workloadrebalancer
+
+import (
+	"context"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/karmada-io/karmada/pkg/apis/apps/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+func (c *RebalancerController) deleteExpiredRebalancer(key util.QueueKey) error {
+	objKey, ok := key.(client.ObjectKey)
+	if !ok {
+		klog.Errorf("Invalid object key: %+v", key)
+		return nil
+	}
+	klog.V(4).Infof("Checking if WorkloadRebalancer(%s) is ready for cleanup", objKey.Name)
+
+	// 1. Get WorkloadRebalancer from cache.
+	rebalancer := &appsv1alpha1.WorkloadRebalancer{}
+	if err := c.Client.Get(context.TODO(), objKey, rebalancer); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// 2. Use the WorkloadRebalancer from cache to see if the TTL expires.
+	if expiredAt, err := c.processTTL(rebalancer); err != nil {
+		return err
+	} else if expiredAt == nil {
+		return nil
+	}
+
+	// 3. The WorkloadRebalancer's TTL is assumed to have expired, but the WorkloadRebalancer TTL might be stale.
+	// Before deleting the WorkloadRebalancer, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
+	fresh := &appsv1alpha1.WorkloadRebalancer{}
+	if err := c.ControlPlaneClient.Get(context.TODO(), objKey, fresh); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// 4. Use the latest WorkloadRebalancer directly from api server to see if the TTL truly expires.
+	if expiredAt, err := c.processTTL(fresh); err != nil {
+		return err
+	} else if expiredAt == nil {
+		return nil
+	}
+
+	// 5. deletes the WorkloadRebalancer if TTL truly expires.
+	options := &client.DeleteOptions{Preconditions: &metav1.Preconditions{ResourceVersion: &fresh.ResourceVersion}}
+	if err := c.ControlPlaneClient.Delete(context.TODO(), fresh, options); err != nil {
+		klog.Errorf("Cleaning up WorkloadRebalancer(%s) failed: %+v", fresh.Name, err)
+		return err
+	}
+	klog.V(4).Infof("Cleaning up WorkloadRebalancer(%s) successful.", fresh.Name)
+
+	return nil
+}
+
+// processTTL checks whether a given WorkloadRebalancer's TTL has expired, and add it to the queue after
+// the TTL is expected to expire if the TTL will expire later.
+func (c *RebalancerController) processTTL(r *appsv1alpha1.WorkloadRebalancer) (expiredAt *time.Time, err error) {
+	if !needsCleanup(r) {
+		return nil, nil
+	}
+
+	remainingTTL, expireAt, err := timeLeft(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// TTL has expired
+	if *remainingTTL <= 0 {
+		return expireAt, nil
+	}
+
+	c.ttlAfterFinishedWorker.AddAfter(client.ObjectKey{Name: r.Name}, *remainingTTL)
+	return nil, nil
+}
+
+func timeLeft(r *appsv1alpha1.WorkloadRebalancer) (*time.Duration, *time.Time, error) {
+	now := time.Now()
+	finishAt := r.Status.LastUpdateTime.Time
```

Comment:
Can you show me where to set the `LastUpdateTime`?

Reply:
Sorry, it is my fault when my last `rebase` process, I encountered many conflict.

When I resolving the conflict, I missed the key line to update `LastUpdateTime`.

https://github.com/karmada-io/karmada/compare/7b02e4e882620de6b4f41fff6d8262c43755b669..32819f2972c056ee53a2a51ebbf0d1f6bf7e7cc9

![image](https://github.com/karmada-io/karmada/assets/30589999/9f805acf-51d9-4da0-af5b-cf2a7bb35ba5)

I will do a quick fix.